incubator/hnc/internal/reconcilers/object.go

```diff
@@ -359,9 +360,20 @@ func (r *ObjectReconciler) syncPropagated(ctx context.Context, log logr.Logger,
 	if srcInst == nil {
 		return actionRemove, nil
 	}
-
 	// If an object doesn't exist, assume it's been deleted or not yet created.
 	exists := inst.GetCreationTimestamp() != v1.Time{}
+	selector := r.getSelector(ctx, log, srcInst)
+
+	instLabels := labels.Set(r.Forest.Get(inst.GetNamespace()).GetLabels())
+	if selector != nil && !selector.Matches(instLabels) {
+		// The object already exists and doesn't need to be updated. This will typically happen when HNC
+		// is restarted - all the propagated objects already exist on the apiserver. Record that it exists
+		// for our statistics.
+		r.recordPropagatedObject(log, inst.GetNamespace(), inst.GetName())
+
+		// Nothing more needs to be done.
+		return actionNop, nil
```

Comment:
The logic seems not right to me. Why do we move this part up? I think it should be that if it doesn't match we will just early exit with `actionNop`. The rest of the logic to `actionWrite` if there's a diff and `actionNop` if there's not should remain the same.

Reply:
@yiqigao217 how much will this logic change with the changes you're making around object overwriting? It may not be worth fine-tuning this function as it exists today if you're about to change the way `syncObject` works.

With that said, I mostly agree with Yiqi. This should be an early-exit, but it shouldn't be a no-op - the object might exist and need to be removed. Your existing code doesn't account for this either - so you should probably add a new test which is to ensure that if you _add_ a label selector to an _existing_ object, that it gets _deleted_ from any namespace in which it no longer belongs.